prov/lnx/src/lnx_ops.c

```diff
@@ -150,29 +110,42 @@ get_rx_entry(struct local_prov_ep *cep, struct iovec *iov, void **desc,
 	/* if lp is NULL, then we don't know where the message is going to
 	 * come from, so allocate the rx_entry from a global pool
 	 */
-	if (!cep) {
+	if (!lep) {
 		bp = global_recv_bp;
 		bplock = &global_bplock;
 	} else {
-		bp = cep->lpe_recv_bp;
-		bplock = &cep->lpe_bplock;
+		bp = lep->le_recv_bp;
+		bplock = &lep->le_bplock;
 	}
 
 	ofi_spin_lock(bplock);
 	rx_entry = (struct lnx_rx_entry *)ofi_buf_alloc(bp);
 	ofi_spin_unlock(bplock);
 	if (rx_entry) {
 		memset(rx_entry, 0, sizeof(*rx_entry));
-		if (!cep)
+		if (!lep)
 			rx_entry->rx_global = true;
-		rx_entry->rx_cep = cep;
+		rx_entry->rx_lep = lep;
 		lnx_init_rx_entry(rx_entry, iov, desc, count, addr, tag,
 				  ignore, context, flags);
 	}
 
 	return rx_entry;
 }
 
+static inline struct lnx_rx_entry *
+lnx_find_first_match(struct lnx_queue *q, struct lnx_match_attr *match)
+{
+	struct lnx_rx_entry *rx_entry;
+
+	ofi_spin_lock(&q->lq_qlock);
```

Comment:
Do we need locks in the queue? Shouldn't the entire peer progress call be locked already?

Reply:
sorry I assumed you locked the entire cq during progress so that progress in each peer was already locked. in the man page we specify that the calls to get_tag/queue_tag need to be serialized, implying that we need a lock around the entire peer progress call.
how does this protect against two threads calling progress at the same time? could it potentially result in two incoming messages completing out of order?